sdks/java/core/src/test/java/org/apache/beam/sdk/io/WriteTest.java

```diff
@@ -129,6 +139,71 @@ public void testWrite() {
   }
 
   /**
+   * Test that Write with an empty input still produces one shard.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testEmptyWrite() {
+    runWrite(Collections.<String>emptyList(), IDENTITY_MAP);
+    // Note we did not request a sharded write, so runWrite will not validate the number of shards.
+    assertEquals(1, numShards.intValue());
+  }
+
+  /**
+   * Test that Write with a configured number of shards produces the desired number of shards even
+   * when there are many elements.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testShardedWrite() {
+    runShardedWrite(
+        Arrays.asList("one", "two", "three", "four", "five", "six"),
+        IDENTITY_MAP,
+        Optional.of(1));
+  }
+
+  /**
+   * Test that Write with a configured number of shards produces the desired number of shards even
+   * when there are too few elements.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testExpandShardedWrite() {
+    runShardedWrite(
+        Arrays.asList("one", "two", "three", "four", "five", "six"),
+        IDENTITY_MAP,
+        Optional.of(20));
+  }
+
+  /**
+   * Tests that a Write can balance many elements.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testShardedWriteBalanced() {
+    int numElements = 1000;
+    List<String> inputs = new ArrayList<>(numElements);
+    for (int i = 0; i < numElements; ++i) {
+      inputs.add(String.format("elt%04d", i));
+    }
+
+    runShardedWrite(
+        inputs,
+        new WindowAndReshuffle<>(
+            Window.<String>into(Sessions.withGapDuration(Duration.millis(1)))),
+        Optional.of(10));
+
+    // Check that both the min and max number of results per shard are close to the expected.
```

Comment:
The test is `min > 0.9*max`. This tests that min is within 10% of max, so it tests both.